plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -1292,6 +1302,85 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
+            throw new TrinoException(NOT_SUPPORTED, "Changing partition column types is not supported");
+        }
+        if (!canChangeColumnType(column.getType(), type)) {
+            throw new TrinoException(NOT_SUPPORTED, "Cannot change type from %s to %s".formatted(column.getType(), type));
+        }
+        metastore.setColumnType(table.getSchemaName(), table.getTableName(), column.getName(), toHiveType(type));
+    }
+
+    private static boolean canChangeColumnType(Type sourceType, Type newType)
+    {
+        if (sourceType.equals(newType)) {
+            return true;
+        }
+        if (sourceType == TINYINT) {
+            return newType == SMALLINT || newType == INTEGER || newType == BIGINT;
+        }
+        if (sourceType == SMALLINT) {
+            return newType == INTEGER || newType == BIGINT;
+        }
+        if (sourceType == INTEGER) {
+            return newType == BIGINT;
+        }
+        if (sourceType == REAL) {
+            return newType == DOUBLE;
+        }
+        if (sourceType instanceof VarcharType || sourceType instanceof CharType) {
+            return newType instanceof VarcharType || newType instanceof CharType;
+        }
+        if (sourceType instanceof DecimalType sourceDecimal && newType instanceof DecimalType newDecimal) {
+            return sourceDecimal.getScale() == newDecimal.getScale()
+                    && sourceDecimal.getPrecision() <= newDecimal.getPrecision();
+        }
+        if (sourceType instanceof ArrayType sourceArrayType && newType instanceof ArrayType newArrayType) {
+            return canChangeColumnType(sourceArrayType.getElementType(), newArrayType.getElementType());
+        }
+        if (sourceType instanceof RowType sourceRowType && newType instanceof RowType newRowType) {
+            List<RowType.Field> fields = Streams.concat(sourceRowType.getFields().stream(), newRowType.getFields().stream())
+                    .distinct()
+                    .collect(toImmutableList());
+            for (RowType.Field field : fields) {
+                String fieldName = field.getName().orElseThrow();
+                if (fieldExists(sourceRowType, fieldName) && fieldExists(newRowType, fieldName)) {
+                    if (!canChangeColumnType(
+                            findFieldByName(sourceRowType.getFields(), fieldName).getType(),
+                            findFieldByName(newRowType.getFields(), fieldName).getType())) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private static RowType.Field findFieldByName(List<RowType.Field> fields, String fieldName)
+    {
+        return fields.stream()
+                .filter(field -> field.getName().orElseThrow().equals(fieldName))
+                .collect(onlyElement());
+    }
+
+    private static boolean fieldExists(RowType structType, String fieldName)
```

Comment:
You iterate over fields twice. Also, one method does this for a stream, while other with a for loop, so it's less obvious they do exact same thing

what about making `findFieldByName` return Optional?
then you don't need `fieldExists` at all

```java
boolean allowedChange = findFieldByName(a, fieldName).flatMap(sourceField ->
   findFieldByName(b, fieldName).map(targetField ->
      canChangeColumnType(....)))
// Allow removal and addition of fields
.orElse(TRUE);
if (!allowedChange) {
  return false;
}
```